tests/src/core/testqgsexpression.cpp

```diff
@@ -2443,6 +2443,43 @@ class TestQgsExpression: public QObject
       QCOMPARE( badArray.evalErrorString(), QString( "Cannot convert 'not an array' to array" ) );
     }
 
+    void compare_arrays()
+    {
+      QCOMPARE( QgsExpression( "array() = array()" ).evaluate(), QVariant( true ) );
+      QCOMPARE( QgsExpression( "array(NULL) = array(NULL)" ).evaluate(), QVariant( true ) );
```

Comment:
That's what I've learnt in oracle DB courses 15 years ago: `NULL` is not equal to `NULL`.